### PR TITLE
proto: Switch to using buf

### PR DESCRIPTION
**Pull request by krishicks**
This replaces `protoc` with `buf generate`. This uses [buf](https://buf.build/).

A quick overview of the benefits of using `buf` can be seen [here](https://docs.buf.build/#the-buf-cli).

I don't think we necessarily need to switch to using it now, but it does make it easier to e.g. add new plugins, and we could take advantage of the linting and breaking-change detector immediately. It also seems to generate slightly more correct file descriptors; see the changes in the `buf generate` commit.

### CODE REVIEW

File: buf.yaml
Comment (schmichael) [inline buf.yaml:22]: While `buf` looks pretty exciting, it seems like we're not actually able to take advantage of much of it. It incurs more configuration overhead and introduces another build dependency.

Are the remaining linters impactful enough to outweigh the costs?

Will future protobufs in Nomad be able to take advantage of more of `buf`s features? If that's the case it'd be nice to see them so we can get a better idea of `buf`s impact.

Are other teams at HashiCorp using `buf`?

Comment (krishicks) [reply buf.yaml:22]: I don't have answers to these questions, so I will just close the PR.

Thanks for reviewing!

File: scripts/vagrant-linux-priv-buf.sh
Comment (bufdev) [inline scripts/vagrant-linux-priv-buf.sh:7]: Note for your usage here, you don't need to download the entire tarball, you can just download the binary (drop the `.tar.gz` and you're good to go.

File: buf.yaml
Comment (bufdev) [inline buf.yaml:9]: Was there a reason you chose this list? The best to start with is `DEFAULT`, and go from there. Note that `buf check lint --error-format=config-ignore-yaml` will generate a lint ignores section you can copy/paste to ignore your current errors on a per-file level.

File: buf.gen.yaml
Comment (bufdev) [inline buf.gen.yaml:10]: We'll have a more generic solution to this problem soon :-)

File: GNUmakefile
Comment (bufdev) [inline GNUmakefile:250]: This can actually be problematic - it's an edge case, but `protoc` (at least in the past, they might have fixed this by now but unlikely) had different results if you compiled per-file as opposed to all the files in one directory. Generally, you'd want to do something along the lines of:

```bash
for dir in $(find . -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq); do
  protoc -I . -o /dev/null $(find "${dir}" -name '*.proto')
done
```

Comment (bufdev) [reply GNUmakefile:250]: Of course, we at Buf have thought about this, so you don't have to worry about this problem with `buf generate` :-)

File: scripts/vagrant-linux-priv-buf.sh
Comment (krishicks) [reply scripts/vagrant-linux-priv-buf.sh:7]: I chose the tarball to get `protoc-gen-buf-check-breaking` and `protoc-gen-buf-check-lint`; my understanding was those binaries needed to exist on `$PATH` to do e.g. `buf check lint` in CircleCI in a later commit.

File: buf.yaml
Comment (krishicks) [reply buf.yaml:9]: This list is a combination of something I copied from the docs as well as the output of the command you gave.

File: buf.gen.yaml
Comment (krishicks) [reply buf.gen.yaml:10]: That's great! I only happened to know about this `M` flag business because of `prototool` adding it when generating protos for another project.

File: buf.yaml
Comment (bufdev) [reply buf.yaml:9]: Ah - that list is just meant for the tour, it's used for googleapis as they have a lot of things that don't conform (semi-obviously). We recommend starting with:

```yaml
lint:
  use:
    - DEFAULT
```

And then going from there.

You can see all available categories and checkers with `buf check ls-lint-checkers --all` as well. The introduction to https://docs.buf.build/lint-checkers gives a rough overview of the hierarchy.

File: scripts/vagrant-linux-priv-buf.sh
Comment (bufdev) [reply scripts/vagrant-linux-priv-buf.sh:7]: Those are not needed for `buf check lint`, those are split out so you can do i.e. `protoc -I . --buf-check-lint_out=.`. All of their functionality is contained within `buf` - what part of the documentation led you to this? We should update it if this was confusing, our apologies.

File: GNUmakefile
Comment (krishicks) [reply GNUmakefile:250]: This is similar to what we ended up doing on another internal HashiCorp repo while beginning the process of switching away from `prototool`. Because we were using `prototool` with 4 plugins and multiple proto roots, the actual implementation was far more elaborate.

I actually started with the snippet you have above but it ended up boiling down to this:

```bash
    for protoDir in $protoDirs; do
      protoFiles=$(find "$protoDir" -name '*.proto' | sort | tr '\n' ' ')

      for plugin in $plugins; do
        cmd="$protocPath $includes $plugin $protoFiles"
        echo "$cmd"
        $cmd
      done
    done
```

Where the `$includes` and `$plugin` are all the wild `-I` flags and plugin-specific invocation+flags for each plugin we use.

Comment (bufdev) [reply GNUmakefile:250]: Sounds about right :-)

File: scripts/vagrant-linux-priv-buf.sh
Comment (krishicks) [reply scripts/vagrant-linux-priv-buf.sh:7]: I think it was by going through the [tour](https://docs.buf.build/tour-1), which says:

> This tour assumes you have installed the buf binary on your $PATH. The protoc plugin section assumes you have also installed the protoc-gen-buf-check-breaking and protoc-gen-buf-check-lint binaries, and have a working installation of protoc.

I didn't get as far as https://docs.buf.build/tour-9, which makes it more clear what the "protoc plugin section" means.

My mistake!

Comment (bufdev) [reply scripts/vagrant-linux-priv-buf.sh:7]: We'll update the docs to clarify

Comment (bufdev) [reply scripts/vagrant-linux-priv-buf.sh:7]: Docs have been updated to not reference the protoc plugin installation until the section of the tour where they are required!

Comment (krishicks) [reply scripts/vagrant-linux-priv-buf.sh:7]: Looks great! Thanks!

File: .circleci/config/commands/install-buf.yml
Comment (bufdev) [inline .circleci/config/commands/install-buf.yml:5]: (You might be on this but) you need to change this too I believe

**Comment by krishicks** [comment]
This PR was updated to include proto breaking change detection, a feature that `buf` provides.

File: GNUmakefile
Comment (tgross) [inline GNUmakefile:206]: We should add the installation of `buf` to the `deps` target and the go.mod file in the `tools/` directory as well, so that we're pinning the version for folks who aren't using Vagrant.

File: buf.gen.yaml
Comment (tgross) [inline buf.gen.yaml:1]: This is totally an aesthetic nitpick but is there any way to get these config files out of the root of the project to reduce the amount of config pollution there?

Comment (krishicks) [reply buf.gen.yaml:1]: It is possible to tell `buf` where to get the files via the `--template` and `--config` flags; I can update it. What's a reasonable subdirectory for this? `.buf`?

Comment (tgross) [reply buf.gen.yaml:1]: That works. Or somewhere in the existing `tools/` dir

File: GNUmakefile
Comment (notnoop) [inline GNUmakefile:207]: I'm curious how this works in practice.  Given that latest master is an ever moving target and isn't always the PR base, we may get false positives or build indeterminism. I may suggest using latest stable release, and update it whenever we cut a release.

Few scenarios I'm thinking about:

* if branch is a release backport based on earlier release, if master contains a new field not found in brach, will the linter complain about [a field deletion without reservation](https://docs.buf.build/breaking-checkers/#field_no_delete_unless_number_reserved). The same scenario may happen with actively developed branches that branched off before a new field got merged into master.

* If we iterate on endpoints that hasn't been released, the linter warnings aren't actionable; relatedly, if we accidentally merged a PR with a breaking change, we'll lose the signal unless we go back and examine the branch.  This shouldn't happen, but I've seen cases where failing tests got merged for variety of legitimate reasons or oversight.

**Review by notnoop — COMMENTED** [review]
This looks great - and buf seems like a nice addition. Echo Tim's question and I have a question about backward compatibility linter. Thanks for introducing us to buf!

File: GNUmakefile
Comment (bufdev) [reply GNUmakefile:207]: Glad it looks like it could be helpful - let us know if you have any questions :-)

On your points:

- Checking against a release is likely the best idea overall, we just document as `master` - note that Buf can take inputs from a variety of endpoints, not just your local git branch https://docs.buf.build/inputs
- This `reserved` breaking change checks are some of the more controversial checks, however to make sure breaking change detection is transitive (and to make sure future commits don't include fields that were used in the past but deleted), we chose to enforce this, BUT it's just as easy to disable this using `breaking.except` https://docs.buf.build/breaking-configuration

Comment (notnoop) [reply GNUmakefile:207]: Makes sense - Thank you @bufdev for the insight and feedback!

File: GNUmakefile
Comment (krishicks) [reply GNUmakefile:206]: I've added buf to the tools, but to avoid a merge conflict I've skipped adding it to the current `deps` make target. This should be pulled in after https://github.com/hashicorp/nomad/pull/9357.

**Comment by krishicks** [comment]
The lint check is failing because for some reason `make deps lint-deps` is installing buf `v0.3.0`. I'm going to wait for the deps PR to be merged before rebasing and pushing again.

File: tools/go.mod
Comment (bufdev) [inline tools/go.mod:10]: Note that separately, this is a change you're going to want to extensively test - while advertised as compatible, there were major changes, including breaking changes on edge cases, from golang/protobuf 1.3.5 to 1.4.0. Version 1.4.0 was re-implemented in terms of v2 github.com/protocolbuffers/protobuf-go, but while advertised as providing compatibility, this is actually only per their [compatibility guarantee](https://github.com/golang/protobuf#compatibility), which makes exceptions, and there have been numerous documented cases of underlying behavior changing in a way that has broken users (do a search on their github issues).

You'll need to upgrade eventually, so you might as well do it now, and if you're not doing anything overly complicated with Protobuf, you might be fine, just a heads up!

Comment (tgross) [reply tools/go.mod:10]: @krishicks the spot this worries me with is with external CSI plugins... we don't own the protobuf, we can't extensively test these because there's 100+ of them and many require specific vendor hardware, and we see all kinds of specification compliance issues with them. If we're going to merge this we should either do it and land the fix in a beta for 1.0 or wait until _after_ 1.0-final ships so that we have time to test at least the simple "known good" cloud vendor plugins.

Comment (krishicks) [reply tools/go.mod:10]: I noticed this, but on recompiling the protos found no change. But thanks for the extra info! I _have_ noticed diffs when upgrading e.g. grpc-gateway that gave me enough pause that I decided to roll back the upgrade in other projects.

File: GNUmakefile
Comment (krishicks) [reply GNUmakefile]: I've updated the `against` to be v1.0.0-beta3; I figure when 1.0.0 is GA we'll change it to that. Seem reasonable?

File: tools/go.mod
Comment (bufdev) [reply tools/go.mod:10]: A separate recommendation: I wouldn't track tool dependencies in your `go.mod` at all - I think this is why it can lead to problems. We track them via `make`, for example https://github.com/bufbuild/buf/blob/master/make/go/dep_golangci_lint.mk. This is likely way over-engineered, but the main concept:

```make
FOO_VERSION := v1.2.3
cd /tmp; GOBIN=/path/to/tools/bin GO111MODULE=on go get github.com/foo/foo@$(FOO_VERSION)
```

Doing the `cd /tmp` avoids `go` touching your `go.mod`. Go 1.16 allow this workflow via `go install` https://tip.golang.org/doc/go1.16#go-command

This means that you can have your code dependencies stay however you want (for example, staying on 1.3.4 for golang/protobuf), and install tools that may have upgraded past this themselves (I spent about 10 hours solving all the issues a few months ago, although granted bufbuild/buf has some intricate usage of golang/protobuf).

Comment (krishicks) [reply tools/go.mod:10]: (My above comment was directed at bufdev)

@tgross The upgrade of protobuf here is because of adding `buf` to tools. It wasn't the case that `protoc` was installed via tools before; so if there's concern about this I think we should _not_ install buf via `make deps` and continue to point at the instructions in the README, which were updated to point at buf rather than protoc.

Comment (tgross) [reply tools/go.mod:10]: > Go 1.16 allow this workflow via go install https://tip.golang.org/doc/go1.16#go-command

We're pinned to go1.15 but once we upgrade to go1.16 that seems like reasonable advice. Not for this PR though.

> @tgross The upgrade of protobuf here is because of adding buf to tools. It wasn't the case that protoc was installed via tools before; so if there's concern about this I think we should not install buf via make deps and continue to point at the instructions in the README, which were updated to point at buf rather than protoc.

Sorry, I might be missing something... shouldn't our README instructions say to use `make deps` so that the installation instructions match exactly what CI will do? And is buf forcing an upgrade to our protobufs version for some reason?

Comment (bufdev) [reply tools/go.mod:10]: Note Go 1.16 isn’t due to come out until February, so using it for this workflow is future-looking advice :-)

Comment (krishicks) [reply tools/go.mod]: I've reverted the change that adds `buf` to tools to avoid the upgrade of golang/protobuf.

**Review by tgross — APPROVED** [review]
LGTM! 👍

**Review by schmichael — APPROVED** [review]
A brief README.md in `tools/buf` would be nice for people who find themselves there out of curiosity or a build error or something. A 1-or-2-liner "what is it and why do we use it" + link to the upstream homepage seems sufficient.

Thanks for the great tool and help @bufdev!

**Comment by bufdev** [comment]
No problem! Let us know if you run into any issues down the line, we're building new stuff every day :-)

**Pull request merged**
